File: cpp/1510_StoneGameIV.cpp
class Solution {
public:
    bool winnerSquareGame(int n) {
        vector<bool> dp = {false};
        vector<int> squares = {1};
        for (int i = 1; i <= n; ++i)  {
            if (sqrt(i) == (int) sqrt(i)) {
                squares.push_back(i);
                dp.push_back(true);
            } else {
                for (int sq : squares) {
                    if (dp[i-sq] == false) {
                        dp.push_back(true);
                        break;
                    }
                }
                if (dp.size() == i) dp.push_back(false);
            }
        }
        return dp[n];
    }
};

File: swift/1510_StoneGameIV.swift
class Solution {
    func winnerSquareGame(_ n: Int) -> Bool {
        var dp = [false]
        var squares = [1]
        for i in 1...n {
            var root = Double(i).squareRoot()
            if root == floor(root) {
                dp.append(true)
                squares.append(i)
                continue
            }
            for square in squares {
                if dp[i-square] == false {
                    dp.append(true)
                    break
                }
            }
            if dp.count == i {
                dp.append(false)
            }
        }
        return dp[n]
    }
}
